Opaque (reference) access token guidance (#36588)
dotnet/AspNetCore.Docs: aspnetcore/blazor/security/additional-scenarios.md

```

:::moniker-end

## Opaque (reference) access token support

> *The following guidance requires an authentication server that supports opaque (reference) access tokens. Currently, Microsoft Entra doesn't support opaque access token validation.*

Reviewer: Based on testing the sample against Keycloak:

1. Keycloak and Okta also issue **JWT** access tokens by default — there isn't a first-class "opaque" toggle. The handler still works against them (introspection doesn't care whether the token is opaque or JWT), but readers reaching for this section because they're using Keycloak/Okta should know that "opaque" here refers to how the *client* treats the token, not how the *server* mints it. Duende IdentityServer is the main implementation in the .NET ecosystem that issues truly opaque reference tokens out of the box.

2. **Keycloak-specific footgun, worth a sentence:** when testing introspection against Keycloak, the API's introspection client must be different from the OIDC client that issued the user's access token. If they're the same client, Keycloak returns `{"active": false}` with `"Access token JWT check failed"` in the server log, and the reader will think the handler is broken. In the documented Blazor Web App scenario this doesn't happen naturally (`BlazorWebAppOidc` and `MinimalApiJwt` are separate clients), but a one-liner saves the next person an hour.

Also, italicized prose for a caveat reads a bit awkwardly here. Consider a NOTE block so it stands out the same way the IMPORTANT block below does::

```suggestion
> [!NOTE]
> The following guidance requires an authentication server that supports opaque (reference) access tokens. Currently, Microsoft Entra doesn't support opaque access token validation. Keycloak and Okta issue JWT access tokens by default; the handler still works against them because it relies only on RFC 7662 introspection, but "opaque" in this section describes how the client treats the token rather than how the server mints it. Duende IdentityServer issues true opaque reference tokens out of the box.
>
> When testing this pattern against Keycloak specifically, the API's introspection client must be a different OIDC client than the one that issued the user's access token. Introspecting a token using the client that minted it returns `{"active": false}`.
```

```dotnetcli
dotnet user-secrets set "Authentication:Schemes:AuthServer:ClientSecret" "{SECRET}"
```

If using Visual Studio, you can confirm the secret is set by right-clicking the server project in **Solution Explorer** and selecting **Manage User Secrets**.

`Extensions/HttpContextExtensions.cs`:

```csharp
namespace MinimalApiJwt.Extensions;

public static class HttpContextExtensions
{
    public static string? ExtractBearerToken(this HttpRequest request)
    {
        var authorizationHeader = request.Headers["Authorization"].ToString();

        if (!string.IsNullOrEmpty(authorizationHeader) && 
            authorizationHeader.StartsWith("Bearer ", 
            StringComparison.OrdinalIgnoreCase))
        {
            var token = authorizationHeader["Bearer ".Length..].Trim();

            if (!string.IsNullOrEmpty(token))
            {
                return token;
            }
        }

        return null;
    }
}
```

`Authentication/OpaqueTokenAuthenticationOptions.cs`:

```csharp
using Microsoft.AspNetCore.Authentication;

namespace MinimalApiJwt.Authentication;

> 
> public class OpaqueTokenAuthenticationOptions : AuthenticationSchemeOptions
> {
>     public const string DefaultScheme = "OpaqueTokenAuthentication";
>     public string? IntrospectionEndpoint { get; set; }
>     public string? ClientId { get; set; }
> }

Reviewer: Related to the `Authentication:Schemes:AuthServer:ClientSecret` comment further down: the cleaner pattern is to add `ClientSecret` to `OpaqueTokenAuthenticationOptions` and let the framework's normal `Authentication:Schemes:{SchemeName}:*` binding pick it up. Then the handler doesn't need an `IConfiguration` dependency at all and the config key in the docs and the `dotnet user-secrets set` command would automatically match the scheme name. Optional, but worth considering since it eliminates the naming-inconsistency surface area entirely.

```suggestion
public class OpaqueTokenAuthenticationOptions : AuthenticationSchemeOptions
{
    public const string DefaultScheme = "OpaqueTokenAuthentication";
    public string? IntrospectionEndpoint { get; set; }
    public string? ClientId { get; set; }
    public string? ClientSecret { get; set; }
}
```

If you apply this, also:
- Drop the `IConfiguration config` parameter from the handler's constructor.
- Replace `config["Authentication:Schemes:AuthServer:ClientSecret"]` with `Options.ClientSecret`.
- Update the `dotnet user-secrets set` command to use `Authentication:Schemes:OpaqueTokenAuthentication:ClientSecret`.

    ILoggerFactory logger,
    UrlEncoder encoder,
    IConfiguration config,
    IHttpClientFactory httpClientFactory)
    : AuthenticationHandler<OpaqueTokenAuthenticationOptions>(options, logger, encoder)
{
    protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
    {
        var opaqueToken = Request.ExtractBearerToken();

        if (opaqueToken is null)
        {
            var failedResult = AuthenticateResult.Fail(
                "Bearer token not found in Authorization header.");
            return failedResult;
        }

        /*
            The following example attempts to validate the opaque 
            (reference) access token.

            An HTTP call is made to the authorization server's introspection 
            endpoint with the token and the API's credentials. The response
            is processed to determine if the token is valid.

            If the token is valid, an AuthenticationTicket is created 
            containing the user's claims.

            If the token is invalid, a failed authorization result is 
            returned.
        */

        var introspectionUri = options.IntrospectionEndpoint;

>         var clientId = options.ClientId;

Reviewer: This won't compile. `options` is the `IOptionsMonitor<OpaqueTokenAuthenticationOptions>` constructor parameter, which doesn't have `IntrospectionEndpoint`/`ClientId` properties. Use the base class's `Options` property (capital `O`) instead, which exposes the resolved options for the current scheme:

```suggestion
        var introspectionUri = Options.IntrospectionEndpoint;
        var clientId = Options.ClientId;
```

If you keep this fix, the `options` constructor parameter is no longer referenced directly and the `using Microsoft.Extensions.Options;` line can stay (still needed for the generic constraint passed to the base type).

>         var clientSecret = config["Authentication:Schemes:AuthServer:ClientSecret"];

Reviewer: The scheme name in this config key (`AuthServer`) doesn't match `OpaqueTokenAuthenticationOptions.DefaultScheme` (`"OpaqueTokenAuthentication"`). The standard `Authentication:Schemes:{SchemeName}:...` convention would make this:

```suggestion
        var clientSecret = config["Authentication:Schemes:OpaqueTokenAuthentication:ClientSecret"];
```

If you change this, also update the `dotnet user-secrets set` command on line 1095 and the explanatory text on line 1084 to use the same key.

        // Prepare the form-encoded body containing the token
        var content = new FormUrlEncodedContent(
        [
            new KeyValuePair<string, string>("token", opaqueToken)
            // NOTE: Some servers require "token_type_hint", for example
            // set to "access_token"
        ]);

        // Post to the introspection endpoint
        var response = await client.PostAsync(introspectionUri, content);

        if (!response.IsSuccessStatusCode)
        {
            var failedResult = AuthenticateResult.Fail(
                "Introspection endpoint failure.");

            return failedResult;
        }

        // Parse the JSON response
        var responseString = await response.Content.ReadAsStringAsync();
        using var doc = JsonDocument.Parse(responseString);

        // The 'active' property determines if the token is valid and not expired
        var tokenIsValid = doc.RootElement.GetProperty("active").GetBoolean();

        if (tokenIsValid)
        {
            // TODO: Replace the '{USER ID}' placeholder with extracted claim value
            // from the token introspection response

>             var claims = new[] { new Claim(ClaimTypes.Name, "{USER ID}") };

Reviewer: **This is the most important issue with the section.** I stood up Keycloak and ran the sample end-to-end with a real access token. With the handler as written, the call succeeds (`IsAuthenticated == true`) but produces a `ClaimsPrincipal` with `Name = "{USER ID}"`, no `NameIdentifier`, no email, and no roles. The `// TODO` is buried inside a code block and very easy to miss — a reader who copies this verbatim ships authentication that *looks* like it works but exposes no usable user information.

Please replace the placeholder with a real mapping from the standard RFC 7662 introspection fields. Suggested:

```suggestion
            // Map standard introspection response fields onto claims.
            // Field names below match what Keycloak, Duende IdentityServer,
            // Auth0, and Okta return; adjust the role source for your provider.
            var claims = new List<Claim>();

            string? Get(string name) =>
                doc.RootElement.TryGetProperty(name, out var v) &&
                v.ValueKind == JsonValueKind.String ? v.GetString() : null;

            var sub = Get("sub");
            var username = Get("preferred_username") ?? Get("username") ?? sub;

            if (sub is not null) claims.Add(new Claim(ClaimTypes.NameIdentifier, sub));
            if (username is not null) claims.Add(new Claim(ClaimTypes.Name, username));
            if (Get("email") is { } email) claims.Add(new Claim(ClaimTypes.Email, email));
            if ((Get("client_id") ?? Get("azp")) is { } cid)
                claims.Add(new Claim("client_id", cid));
            if (Get("scope") is { } scope)
                foreach (var s in scope.Split(' ', StringSplitOptions.RemoveEmptyEntries))
                    claims.Add(new Claim("scope", s));

            // Keycloak surfaces realm roles under realm_access.roles.
            // Duende/IdentityServer uses a flat "role" claim; Auth0 uses a
            // configurable custom claim. Adjust for your authorization server.
            if (doc.RootElement.TryGetProperty("realm_access", out var ra) &&
                ra.ValueKind == JsonValueKind.Object &&
                ra.TryGetProperty("roles", out var roles) &&
                roles.ValueKind == JsonValueKind.Array)
            {
                foreach (var r in roles.EnumerateArray())
                    if (r.ValueKind == JsonValueKind.String)
                        claims.Add(new Claim(ClaimTypes.Role, r.GetString()!));
            }

            var identity = new ClaimsIdentity(claims,
                OpaqueTokenAuthenticationOptions.DefaultScheme,
                nameType: ClaimTypes.Name,
                roleType: ClaimTypes.Role);
```

Side-by-side test result against the same Keycloak token (user `alice` in roles `reader`, `writer`):

| | Doc handler (verbatim) | With the mapping above |
|---|---|---|
| `User.Identity.Name` | `"{USER ID}"` | `"alice"` |
| `User.IsInRole("reader")` | `false` | `true` |
| Claim count | 1 | 9 |

            var identity = new ClaimsIdentity(claims,
                OpaqueTokenAuthenticationOptions.DefaultScheme);
            var principal = new ClaimsPrincipal(identity);
            var ticket = new AuthenticationTicket(principal,
                OpaqueTokenAuthenticationOptions.DefaultScheme);

            var result = AuthenticateResult.Success(ticket);

            return result;
        }
        else
        {
            var failedResult = AuthenticateResult.Fail("Bearer token invalid.");

            return failedResult;
        }
    }
}
```

In the `Program` file:

```csharp
builder.Services.AddHttpClient();
builder.Services.AddAuthentication()
    .AddScheme<OpaqueTokenAuthenticationOptions, OpaqueTokenAuthenticationHandler>(
        OpaqueTokenAuthenticationOptions.DefaultScheme,
        options =>
        {
            options.IntrospectionEndpoint = "{AUTH SERVER INTROSPECTION URI}";
            options.ClientId = "{API CLIENT ID}";
        });
```

The preceding example's placeholders:

* `{AUTH SERVER INTROSPECTION URI}`: Authentication server's introspection URI

> * `{API CLIENT ID}`: API Client ID

Reviewer: Nit. Sentence case per the style guide:

```suggestion
* `{API CLIENT ID}`: API client ID
```

The following <xref:Microsoft.AspNetCore.Authentication.AuthenticationHandler%601> and associated configuration and helper code is provided as a general approach, which might require further development to suit a specific authorization server's requirements. The following handler extracts the opaque token from the `Authorization` header for an HTTP call to an authorization server's introspection endpoint and creates an <xref:Microsoft.AspNetCore.Authentication.AuthenticationTicket> containing the user's claims.

Calling an authorization server's introspection endpoint requires authentication. The following example relies on setting the client secret for authentication in the request's Authorization header (base64 encoded credentials) using the [Secret Manager tool](xref:security/app-secrets) for local development and testing.

[!INCLUDE[](~/blazor/security/includes/secure-authentication-flows.md)]

In the following handler, the authorization server's introspection endpoint client secret uses the configuration key `Authentication:Schemes:OpaqueTokenAuthentication:ClientSecret`.

> 

Reviewer 2: Maybe add a note here: client assertions are recommended in production (Instead of client "shared" secrets)

Author: With what details/cross-link do you recommend? Is the MSAL doc a reasonable?

https://learn.microsoft.com/entra/msal/dotnet/acquiring-tokens/web-apps-apis/confidential-client-assertions

I'm considering the following for this paragraph ...

> In the following handler, the authorization server's introspection endpoint client secret uses the configuration key \`Authentication:Schemes:OpaqueTokenAuthentication:ClientSecret\`. For production apps, consider using \*client assertions\*. For more information, see \[Confidential client assertions (Microsoft Entra documentation)](/entra/msal/dotnet/acquiring-tokens/web-apps-apis/confidential-client-assertions).

Author: I added the remark on the last commit.

>
> When testing this pattern against Keycloak, the API's introspection client must be a different OIDC client than the one that issued the user's access token. Introspecting a token using the client that minted it returns `{"active": false}` with "`Access token JWT check failed`" in the server's log. This doesn't happen naturally for the following scenario because the Blazor Web App and the Minimal API (`MinimalApiJwt`) are separate clients.

<xref:Microsoft.Extensions.DependencyInjection.OpenIdConnectExtensions.AddOpenIdConnect%2A> supports opaque tokens because it doesn't perform access token validation when configured for Proof Key for Code Exchange (PKCE) authorization code flow. It relies on the ASP.NET Core server's HTTPS backchannel to the OIDC authentication service to obtain the ID token using the authorization code received when the user redirects back to the ASP.NET Core app after signing in. If the app is only required to log a user in with OIDC to get a valid authentication cookie, opaque access tokens are supported without modifying the app.

> A failure occurs only when the opaque token acquired by <xref:Microsoft.Extensions.DependencyInjection.OpenIdConnectExtensions.AddOpenIdConnect%2A> is passed to another service that attempts to validate it with <xref:Microsoft.Extensions.DependencyInjection.JwtBearerExtensions.AddJwtBearer%2A>. Unlike self-contained JWTs, opaque tokens require a request to an authorization server to validate their status and retrieve claims. To work around this limitation, either use a third-party API, such as the [Duende Introspection Authentication Handler](https://docs.duendesoftware.com/introspection/), or create a [custom `AuthenticationHandler`](xref:security/authentication/index#authentication-handler) to validate the token.

Reviewer 2: Unlike self-contained JWTs, opaque tokens require a request to an authorization server to validate their status and retrieve claims. 

=> Unlike self-contained JWTs, opaque tokens require a request to an authorization server to validate the status and to retrieve the claims.

> A failure occurs only when the opaque token acquired by <xref:Microsoft.Extensions.DependencyInjection.OpenIdConnectExtensions.AddOpenIdConnect%2A> is passed to another service that attempts to validate it with <xref:Microsoft.Extensions.DependencyInjection.JwtBearerExtensions.AddJwtBearer%2A>. Unlike self-contained JWTs, opaque tokens require a request to an authorization server to validate their status and retrieve claims. To work around this limitation, either use a third-party API, such as the [Duende Introspection Authentication Handler](https://docs.duendesoftware.com/introspection/), or create a [custom `AuthenticationHandler`](xref:security/authentication/index#authentication-handler) to validate the token.

Author: ```suggestion
A failure occurs only when the opaque token acquired by <xref:Microsoft.Extensions.DependencyInjection.OpenIdConnectExtensions.AddOpenIdConnect%2A> is passed to another service that attempts to validate it with <xref:Microsoft.Extensions.DependencyInjection.JwtBearerExtensions.AddJwtBearer%2A>. Unlike self-contained JWTs, opaque tokens require a request to an authorization server to validate the status and to retrieve the claims. To work around this limitation, either use a third-party API, such as the [Duende Introspection Authentication Handler](https://docs.duendesoftware.com/introspection/), or create a [custom `AuthenticationHandler`](xref:security/authentication/index#authentication-handler) to validate the token.
```

            if (doc.RootElement.TryGetProperty("realm_access", out var ra) &&
                ra.ValueKind == JsonValueKind.Object &&
                ra.TryGetProperty("roles", out var roles) &&
                roles.ValueKind == JsonValueKind.Array)
            {
                foreach (var r in roles.EnumerateArray())
                    if (r.ValueKind == JsonValueKind.String)
                        claims.Add(new Claim(ClaimTypes.Role, r.GetString()!));
            }

            var identity = new ClaimsIdentity(claims,
                OpaqueTokenAuthenticationOptions.DefaultScheme,
                nameType: ClaimTypes.Name,
                roleType: ClaimTypes.Role);
            var principal = new ClaimsPrincipal(identity);
            var ticket = new AuthenticationTicket(principal,
                OpaqueTokenAuthenticationOptions.DefaultScheme);

            var result = AuthenticateResult.Success(ticket);

            return result;
        }
        else
        {
            var failedResult = AuthenticateResult.Fail("Bearer token invalid.");

            return failedResult;
        }
    }
}
```

> 

Reviewer 2: Maybe add a note, this code can be further improved by using the OpenID Connect discovery endpoint and adding a cache for the client HttpClient Introspection requests.

> 

Author: ```suggestion

> [!NOTE]
> The preceding approach can be further improved by using the OpenID Connect discovery endpoint and adding a cache for the client's <xref:System.Net.Http.HttpClient> introspection requests.
```

```

:::moniker-end

## Opaque (reference) access token support

> The guidance in this section explains how to implement opaque (reference) access token support, which offers the following advantages over JSON Web Tokens (JWTs):

Reviewer: Italicized prose for a caveat reads a bit awkwardly here — consider a NOTE block so it stands out the same way the IMPORTANT block below does:

```suggestion
> [!NOTE]
> The following guidance requires an authentication server that supports opaque (reference) access tokens. Currently, Microsoft Entra doesn't support opaque access token validation.
```

Author: I'm not sure what you're referring to. This isn't a caveat, and it's no longer italicized. We usually don't start sections with NOTE remarks. It currently begins ...

> The guidance in this section explains how to implement opaque (reference) access token support, which offers the following advantages over JSON Web Tokens (JWTs):
>
> * Strict revocation: Invalidate access tokens at any time before they naturally expire.
> * Token size limits: Store a large number of user claims in the token to avoid a prohibitively large JWT.
> * Security: Prevent API consumers or third parties from reading access token claims.

I felt that that would be a good introduction to the section. Is that a poor introduction?

> The guidance in this section explains how to implement opaque (reference) access token support, which offers the following advantages over JSON Web Tokens (JWTs):

Reviewer: A couple of additions worth folding into this caveat, based on testing the sample against Keycloak:

1. Keycloak and Okta also issue **JWT** access tokens by default — there isn't a first-class "opaque" toggle. The handler still works against them (introspection doesn't care whether the token is opaque or JWT), but readers reaching for this section because they're using Keycloak/Okta should know that "opaque" here refers to how the *client* treats the token, not how the *server* mints it. Duende IdentityServer is the main implementation in the .NET ecosystem that issues truly opaque reference tokens out of the box.

2. **Keycloak-specific footgun, worth a sentence:** when testing introspection against Keycloak, the API's introspection client must be different from the OIDC client that issued the user's access token. If they're the same client, Keycloak returns `{"active": false}` with `"Access token JWT check failed"` in the server log, and the reader will think the handler is broken. In the documented Blazor Web App scenario this doesn't happen naturally (`BlazorWebAppOidc` and `MinimalApiJwt` are separate clients), but a one-liner saves the next person an hour.

Suggested rewrite:

```suggestion
> [!NOTE]
> The following guidance requires an authentication server that supports opaque (reference) access tokens. Currently, Microsoft Entra doesn't support opaque access token validation. Keycloak and Okta issue JWT access tokens by default; the handler still works against them because it relies only on RFC 7662 introspection, but "opaque" in this section describes how the client treats the token rather than how the server mints it. Duende IdentityServer issues true opaque reference tokens out of the box.
>
> When testing this pattern against Keycloak specifically, the API's introspection client must be a different OIDC client than the one that issued the user's access token. Introspecting a token using the client that minted it returns `{"active": false}`.
```

Author: Again, I wonder if you're looking at the latest commits. This was all updated per your prior remarks.

```dotnetcli
dotnet user-secrets init
```

Execute the following command to set the client secret for the authorization server. The `{SECRET}` placeholder is the client secret:

```dotnetcli
dotnet user-secrets set "Authentication:Schemes:OpaqueTokenAuthentication:ClientSecret" "{SECRET}"
```

If using Visual Studio, you can confirm the secret is set by right-clicking the server project in **Solution Explorer** and selecting **Manage User Secrets**.

`Extensions/HttpRequestExtensions.cs`:

```csharp
namespace MinimalApiJwt.Extensions;

public static class HttpRequestExtensions
{
    public static string? ExtractBearerToken(this HttpRequest request)
    {
        var authorizationHeader = request.Headers.Authorization.ToString();

        if (!string.IsNullOrEmpty(authorizationHeader) && 
            authorizationHeader.StartsWith("Bearer ", 
            StringComparison.OrdinalIgnoreCase))
        {
            var token = authorizationHeader["Bearer ".Length..].Trim();

            if (!string.IsNullOrEmpty(token))
            {
                return token;
            }
        }

        return null;
    }
}

> ```
> 
> `Authentication/OpaqueTokenAuthenticationOptions.cs`:
> 
> ```csharp
> using Microsoft.AspNetCore.Authentication;

Reviewer: Related to the `Authentication:Schemes:AuthServer:ClientSecret` comment further down: the cleaner pattern is to add `ClientSecret` to `OpaqueTokenAuthenticationOptions` and let the framework's normal `Authentication:Schemes:{SchemeName}:*` binding pick it up — then the handler doesn't need an `IConfiguration` dependency at all and the config key in the docs and the `dotnet user-secrets set` command would automatically match the scheme name. Optional, but worth considering since it eliminates the naming-inconsistency surface area entirely.

```suggestion
public class OpaqueTokenAuthenticationOptions : AuthenticationSchemeOptions
{
    public const string DefaultScheme = "OpaqueTokenAuthentication";
    public string? IntrospectionEndpoint { get; set; }
    public string? ClientId { get; set; }
    public string? ClientSecret { get; set; }
}
```

If you apply this, also:
- Drop the `IConfiguration config` parameter from the handler's constructor.
- Replace `config["Authentication:Schemes:AuthServer:ClientSecret"]` with `Options.ClientSecret`.
- Update the `dotnet user-secrets set` command to use `Authentication:Schemes:OpaqueTokenAuthentication:ClientSecret`.

Author: Your suggestions were already addressed.

For the request's content in <xref:System.Net.Http.FormUrlEncodedContent>, some servers require a token type hint (`token_type_hint`). For example, the required value might be `access_token`. See your authentication server's documentation for details.

`Authentication/OpaqueTokenAuthenticationHandler.cs`:

```csharp
using System.Net.Http.Headers;
using System.Security.Claims;
using System.Text.Encodings.Web;
using System.Text.Json;
using Microsoft.AspNetCore.Authentication;
using Microsoft.Extensions.Options;
using MinimalApiJwt.Extensions;

namespace MinimalApiJwt.Authentication;

public class OpaqueTokenAuthenticationHandler(
    IOptionsMonitor<OpaqueTokenAuthenticationOptions> optionsMonitor,
    ILoggerFactory logger,
    UrlEncoder encoder,
    IHttpClientFactory httpClientFactory)
    : AuthenticationHandler<OpaqueTokenAuthenticationOptions>(optionsMonitor, 
        logger, encoder)
{
    protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
    {
        var opaqueToken = Request.ExtractBearerToken();

        if (opaqueToken is null)

>         {
>             var failedResult = AuthenticateResult.Fail(

Reviewer: This won't compile. `options` is the `IOptionsMonitor<OpaqueTokenAuthenticationOptions>` constructor parameter, which doesn't have `IntrospectionEndpoint`/`ClientId` properties. Use the base class's `Options` property (capital `O`) instead, which exposes the resolved options for the current scheme:

```suggestion
        var introspectionUri = Options.IntrospectionEndpoint;
        var clientId = Options.ClientId;
```

If you keep this fix, the `options` constructor parameter is no longer referenced directly and the `using Microsoft.Extensions.Options;` line can stay (still needed for the generic constraint passed to the base type).

Author: Already addressed by an earlier commit.

>                 "Bearer token not found in Authorization header.");

Reviewer: The scheme name in this config key (`AuthServer`) doesn't match `OpaqueTokenAuthenticationOptions.DefaultScheme` (`"OpaqueTokenAuthentication"`). The standard `Authentication:Schemes:{SchemeName}:...` convention would make this:

```suggestion
        var clientSecret = config["Authentication:Schemes:OpaqueTokenAuthentication:ClientSecret"];
```

If you change this, also update the `dotnet user-secrets set` command on line 1095 and the explanatory text on line 1084 to use the same key.

Author: Already addressed by an earlier commit.

            string.IsNullOrWhiteSpace(clientId) ||
            string.IsNullOrWhiteSpace(clientSecret))
        {
            var failedResult = AuthenticateResult.Fail(
                "Opaque token authentication isn't fully configured.");
            return failedResult;
        }

        using var client = httpClientFactory.CreateClient();

        // Set the Authorization header (base64 encoded credentials)
        var authString = Convert.ToBase64String(
            System.Text.Encoding.ASCII.GetBytes($"{clientId}:{clientSecret}"));
        client.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue("Basic", authString);

        // Prepare the form-encoded body containing the token
        var content = new FormUrlEncodedContent(
        [
            new KeyValuePair<string, string>("token", opaqueToken)
        ]);

        // Post to the introspection endpoint
        var response = await client.PostAsync(introspectionUri, content);

        if (!response.IsSuccessStatusCode)
        {
            var failedResult = AuthenticateResult.Fail(
                "Introspection endpoint failure.");

> 
>             return failedResult;
>         }

Reviewer: **This is the most important issue with the section.** I stood up Keycloak and ran the sample end-to-end with a real access token. With the handler as written, the call succeeds (`IsAuthenticated == true`) but produces a `ClaimsPrincipal` with `Name = "{USER ID}"`, no `NameIdentifier`, no email, and no roles. The `// TODO` is buried inside a code block and very easy to miss — a reader who copies this verbatim ships authentication that *looks* like it works but exposes no usable user information.

Please replace the placeholder with a real mapping from the standard RFC 7662 introspection fields. Suggested:

```suggestion
            // Map standard introspection response fields onto claims.
            // Field names below match what Keycloak, Duende IdentityServer,
            // Auth0, and Okta return; adjust the role source for your provider.
            var claims = new List<Claim>();

            string? Get(string name) =>
                doc.RootElement.TryGetProperty(name, out var v) &&
                v.ValueKind == JsonValueKind.String ? v.GetString() : null;

            var sub = Get("sub");
            var username = Get("preferred_username") ?? Get("username") ?? sub;

            if (sub is not null) claims.Add(new Claim(ClaimTypes.NameIdentifier, sub));
            if (username is not null) claims.Add(new Claim(ClaimTypes.Name, username));
            if (Get("email") is { } email) claims.Add(new Claim(ClaimTypes.Email, email));
            if ((Get("client_id") ?? Get("azp")) is { } cid)
                claims.Add(new Claim("client_id", cid));
            if (Get("scope") is { } scope)
                foreach (var s in scope.Split(' ', StringSplitOptions.RemoveEmptyEntries))
                    claims.Add(new Claim("scope", s));

            // Keycloak surfaces realm roles under realm_access.roles.
            // Duende/IdentityServer uses a flat "role" claim; Auth0 uses a
            // configurable custom claim. Adjust for your authorization server.
            if (doc.RootElement.TryGetProperty("realm_access", out var ra) &&
                ra.ValueKind == JsonValueKind.Object &&
                ra.TryGetProperty("roles", out var roles) &&
                roles.ValueKind == JsonValueKind.Array)
            {
                foreach (var r in roles.EnumerateArray())
                    if (r.ValueKind == JsonValueKind.String)
                        claims.Add(new Claim(ClaimTypes.Role, r.GetString()!));
            }

            var identity = new ClaimsIdentity(claims,
                OpaqueTokenAuthenticationOptions.DefaultScheme,
                nameType: ClaimTypes.Name,
                roleType: ClaimTypes.Role);
```

Side-by-side test result against the same Keycloak token (user `alice` in roles `reader`, `writer`):

| | Doc handler (verbatim) | With the mapping above |
|---|---|---|
| `User.Identity.Name` | `"{USER ID}"` | `"alice"` |
| `User.IsInRole("reader")` | `false` | `true` |
| Claim count | 1 | 9 |

Author: Already addressed by an earlier commit.

            string? Get(string name) =>
                doc.RootElement.TryGetProperty(name, out var v) &&
                v.ValueKind == JsonValueKind.String ? v.GetString() : null;

            var sub = Get("sub");
            var username = Get("preferred_username") ?? Get("username") ?? sub;

            if (sub is not null) claims.Add(new Claim(ClaimTypes.NameIdentifier, sub));
            if (username is not null) claims.Add(new Claim(ClaimTypes.Name, username));
            if (Get("email") is { } email) claims.Add(new Claim(ClaimTypes.Email, email));
            if ((Get("client_id") ?? Get("azp")) is { } cid)
                claims.Add(new Claim("client_id", cid));
            if (Get("scope") is { } scope)
                foreach (var s in scope.Split(' ', StringSplitOptions.RemoveEmptyEntries))
                    claims.Add(new Claim("scope", s));

            // Keycloak surfaces realm roles under realm_access.roles.
            // Duende/IdentityServer uses a flat "role" claim; Auth0 uses a
            // configurable custom claim. Adjust for your authorization server.
            if (doc.RootElement.TryGetProperty("realm_access", out var ra) &&

>                 ra.ValueKind == JsonValueKind.Object &&

Reviewer: Sentence case per the style guide:

```suggestion
* `{API CLIENT ID}`: API client ID
```

Author: Already addressed by an earlier commit.

            {
                foreach (var r in roles.EnumerateArray())
                    if (r.ValueKind == JsonValueKind.String)
                        claims.Add(new Claim(ClaimTypes.Role, r.GetString()!));
            }

            var identity = new ClaimsIdentity(claims,
                OpaqueTokenAuthenticationOptions.DefaultScheme,
                nameType: ClaimTypes.Name,
                roleType: ClaimTypes.Role);
            var principal = new ClaimsPrincipal(identity);
            var ticket = new AuthenticationTicket(principal,
                OpaqueTokenAuthenticationOptions.DefaultScheme);

            var result = AuthenticateResult.Success(ticket);

            return result;
        }
        else
        {
            var failedResult = AuthenticateResult.Fail("Bearer token invalid.");

            return failedResult;
        }
    }
}
```

> [!NOTE]
> The preceding approach can be further improved by using the OpenID Connect discovery endpoint and adding a cache for the client's <xref:System.Net.Http.HttpClient> introspection requests.
In the `Program` file:

```csharp
using MinimalApiJwt.Authentication;

...

builder.Services.AddHttpClient();

> builder.Services.AddAuthentication()
>     .AddScheme<OpaqueTokenAuthenticationOptions, OpaqueTokenAuthenticationHandler>(
>         OpaqueTokenAuthenticationOptions.DefaultScheme,
>         options =>
>         {
>             options.IntrospectionEndpoint = "{AUTH SERVER INTROSPECTION URI}";
>             options.ClientId = "{API CLIENT ID}";
>         });

Reviewer: As written, this snippet never sets `Options.ClientSecret`, so the handler's new guard clause fails every request with `"Opaque token authentication isn't fully configured."` and returns 401. `AddScheme<TOptions, THandler>` doesn't auto-bind the `Authentication:Schemes:{name}:*` section the way `AddJwtBearer`/`AddOpenIdConnect` do — that auto-binding lives in `JwtBearerConfigureOptions`/`OpenIdConnectConfigureOptions`, not the base API. So readers who only follow the user-secrets step earlier in the article will hit a broken sample.

Verified against Keycloak: with the line below added, alice signs in with all expected claims and roles; without it, every call 401s.

```suggestion
        options =>
        {
            options.IntrospectionEndpoint = "{AUTH SERVER INTROSPECTION URI}";
            options.ClientId = "{API CLIENT ID}";
            options.ClientSecret = builder.Configuration["Authentication:Schemes:OpaqueTokenAuthentication:ClientSecret"];
        });
```

And add `{SECRET}` to the placeholder list below alongside the other two.

Author: Yes, this one needs to be addressed. Got it on my local branch and will commit it shortly.

```

The preceding example's placeholders:

* `{AUTH SERVER INTROSPECTION URI}`: Authentication server's introspection URI
* `{API CLIENT ID}`: API client ID

Values for the authentication server introspection URI (`{AUTH SERVER INTROSPECTION URI}`) and the API client ID (`{API CLIENT ID}`) can be supplied from app settings or any other configuration source.

Tokens are typically invalidated on a logout event using the revocation endpoint. The following example is a starting point for further development:

```csharp
app.MapPost("/logout", 

>     async ([FromForm] string? returnUrl, HttpContext context, 
>     IHttpClientFactory httpClientFactory) =>
> {
>     var accessToken = await context.GetTokenAsync("access_token");
> 
>     if (!string.IsNullOrEmpty(accessToken))
>     {
>         // Prepare the revocation request (RFC 7009)
>         var content = 
>             new FormUrlEncodedContent(new Dictionary<string, string>
>             {
>                 { "token", accessToken },
>                 { "token_type_hint", "access_token" },
>                 { "client_id", "{API CLIENT ID}" },
>                 { "client_secret", "{CLIENT SECRET}" }
>             });
> 
>         // POST to the revocation endpoint
>         using var client = httpClientFactory.CreateClient();
> 
>         await client.PostAsync("{AUTH SERVER TOKEN REVOCATION URI}", content);
>     }
> 
>     TypedResults.SignOut(GetAuthProperties(returnUrl), 
>         [CookieAuthenticationDefaults.AuthenticationScheme]);
> });

Reviewer: Two issues in this snippet:

1. `TypedResults.SignOut(...)` is constructed and discarded — there's no `return`, so the framework never executes it and the user is never actually signed out. The lambda needs to return the result (and stop being `async void`-shaped).
2. `GetAuthProperties` isn't defined or referenced anywhere else in this article. It comes from the BlazorWebAppOidc sample, so a reader landing here from search has no idea where it's from. A one-line callout (or inlining `new AuthenticationProperties { RedirectUri = ... }`) would save them a hunt.

Suggested minimal fix for #1:

```suggestion
app.MapPost("/logout", 
    async ([FromForm] string? returnUrl, HttpContext context, 
    IHttpClientFactory httpClientFactory) =>
{
    var accessToken = await context.GetTokenAsync("access_token");

    if (!string.IsNullOrEmpty(accessToken))
    {
        // Prepare the revocation request (RFC 7009)
        var content = 
            new FormUrlEncodedContent(new Dictionary<string, string>
            {
                { "token", accessToken },
                { "token_type_hint", "access_token" },
                { "client_id", "{API CLIENT ID}" },
                { "client_secret", "{CLIENT SECRET}" }
            });

        // POST to the revocation endpoint
        using var client = httpClientFactory.CreateClient();

        await client.PostAsync("{AUTH SERVER TOKEN REVOCATION URI}", content);
    }

    return TypedResults.SignOut(GetAuthProperties(returnUrl), 
        [CookieAuthenticationDefaults.AuthenticationScheme]);
});
```

Author: Got it ... I'll put on the next commit.
